File: Dotnet don_t delete/Language/2.Inheritance/1.ObjectClassTest/DemoApp/Interval.cs
/*  
    Generally when we use "record" to create a class then we get toString(),
    getHashCode(), Equals() method in this program. Sir has shown one Interval class 
    with the implementation same as the class which we will get without the 
    ==> setter method () == hence they are the immutable object means state/values 
    inside them can't changed.
*/
class Interval
{
    //immutable properties==> we can't change the values through setters
    public int Minutes { get; }
    public int Seconds { get; }

    public Interval(int min, int sec)
    {
        Minutes = min + sec / 60;
        Seconds = sec % 60;
    }

    public int Time()
    {//return o/p in the seconds
        return 60 * Minutes + Seconds;
    }
    //operator overloading
    public static Interval operator+(Interval lhs, Interval rhs)
    {
        return new Interval(lhs.Minutes + rhs.Minutes, lhs.Seconds + rhs.Seconds);
    }

    public override string ToString()
    {
        if(Seconds < 10)
            return Minutes + ":0" + Seconds;
        return Minutes + ":" + Seconds;
    }

    public override int GetHashCode()
    {
        return Minutes + Seconds;
    }

    public override bool Equals(object other)
    {//in Java we are writing the "instanceof" method but in the dotnet we are writing (X is Y that)
        if(other is Interval that)
        {
            //Interval that = (Interval) other;
            return (this.Minutes == that.Minutes) && (this.Seconds == that.Seconds);
        }
        return false;
    }
}

File: Java Don't delete/1.Language/3. Generics/ListTest/Interval.java
class Interval implements Comparable<Interval> {
// instance fields
	private int min;
	private int sec;
//constructor 
	public Interval(int m, int s) {
		min = m + s / 60;
		sec = s % 60; 
	}
//Getters and Setters
	public int minutes() {
		return min;
	}

	public int seconds() {
		return sec;
	}

	public int time() {
		return 60 * min + sec;
	}
// toString ,hashcode  and equals method
	public String toString() {
		if(sec < 10)
			return min + ":0" + sec;
		return min + ":" + sec;
	}

	public int hashCode() {
		return min + sec;
	}
//
	public boolean equals(Object other) {
		if(other instanceof Interval) {
			//if other is the instance of Interval class then downcast "other" to that "object"
			Interval that = (Interval)other; //explicit down-casting
			// to check both the objects are same or not
			return (this.min == that.min) && (this.sec == that.sec);
		}
		return false;
	}

	public int compareTo(Interval that) {
		return this.time() - that.time();
	}
}

